tensorflow_quantum/core/src/circuit_parser.cc

```diff
@@ -159,4 +159,45 @@ Status CircuitFromPauliTerm(const tfq::proto::PauliTerm& term,
   return CircuitFromProgram(measurement_program, num_qubits, circuit);
 }
 
+Status ZBasisCircuitFromPauliTerm(const tfq::proto::PauliTerm& term,
+                                  const int num_qubits, Circuit* circuit) {
+  Program measurement_program;
+  measurement_program.mutable_circuit()->set_scheduling_strategy(
+      cirq::google::api::v2::Circuit::MOMENT_BY_MOMENT);
+  Moment* term_moment = measurement_program.mutable_circuit()->add_moments();
+  float transform_exponent = 0.0;
+  std::string gate_type;
+  for (const tfq::proto::PauliQubitPair& pair : term.paulis()) {
+    if (pair.pauli_type() == "Z") {
+      // Z requires no transform.
+      continue;
+    }
+
+    // Assume it is X and the transform is Y^-0.5
+    transform_exponent = -0.5;
```

Comment:
I like H gate, but Cirq actually has a function called `to_z_basis_ops` that belongs to `cirq.PauliString` which does all basis conversions like this. I thought it would be nice if they were consistent.

I will say this method is also less code because all we have is an EigenGate where we are changing the exponent values as apposed to having to do an HPowGate vs an HPowGate + YPowGate for the S gate.

Reply:
I guess what I am worried about here is that Y^-0.5 does not map the X basis to the Z basis: (Y^-0.5)|+> = (1/sqrt(2))[1-i, 0] =/= [1, 0].  Instead, we have the weaker condition that <+|(Y^-0.5)^\dagger Z (Y^-0.5)|+> = 1.  Similarly for the use of X^0.5.  This weaker condition is what they seem to use in their  `to_z_basis_ops` test [here](https://github.com/quantumlib/Cirq/blob/626e16e41d12098e201ce48a4bb25c047cc86bcd/cirq/ops/pauli_string_test.py#L546).

I can't actually find what they use for their X and Y eigenstate conventions (as seen [here](https://en.wikipedia.org/wiki/Pauli_matrices#Eigenvectors_and_eigenvalues), there are choices), so I'm not even sure that my analysis above is correct, since I am not sure that cirq takes |+x> = (1/sqrt(2))[1, 1].  However, I do find that the matrix of `cirq.X(cirq.GridQubit(0, 0)).to_z_basis_ops()` equals Y^-0.5 and the matrix of `cirq.Y(cirq.GridQubit(0, 0)).to_z_basis_ops()` equals X^0.5, so I agree we are consistent with Cirq.  Do you know where Cirq defines their choice of X and Y eigenvectors?

Reply:
I think we might be doing our math differently, I'm getting that that is all well and good when I check Y^0.5 | +> . Here is a script to double check things that I adopted from the test you linked:
```python
import cirq
import numpy as np

# our computation basis states.
# zero = |0>
# one = |1>
zero = np.array([1, 0])
one =  np.array([0, 1])
print('Z basis dot product=', np.vdot(zero, one))

# our x basis states.
# zero = |0> + |1> 
# one = |0> - |1>
x_zero = np.array([1,1]) / np.sqrt(2)
x_one = np.array([1,-1]) / np.sqrt(2)
print('X basis dot product=', np.vdot(x_zero, x_one))

# our y basis states.
# zero = |0> + i|1>
# one = |0> - i|1>
y_zero = np.array([1, 1j]) / np.sqrt(2)
y_one = np.array([1, -1j]) / np.sqrt(2)
print('Y basis dot product=', np.vdot(y_zero, y_one))

X_trans = (cirq.Y**-0.5)._unitary_()
print('Checking that X transform takes us from x states'
	  ' to computation states.')
print('Zero:', cirq.equal_up_to_global_phase(X_trans @ x_zero, zero))
print('One:', cirq.equal_up_to_global_phase(X_trans @ x_one, one))

Y_trans = (cirq.X**0.5)._unitary_()

print('Checking that Y transform takes us from y states'
	  ' to computation states.')
print('Zero:', cirq.equal_up_to_global_phase(Y_trans @ y_zero, zero))
print('One:', cirq.equal_up_to_global_phase(Y_trans @ y_one, one))
```
Output is
```
Z basis dot product= 0
X basis dot product= 0.0
Y basis dot product= 0j
Checking that X transform takes us from x states to computation states.
Zero: True
One: True
Checking that Y transform takes us from y states to computation states.
Zero: True
One: True
```
Does this clear things up at all ?

Reply:
Added the following multiqubit test:

`# Now, run a comparison against multiqubit states.
# Our computational basis states.
# zero = |00>
# one = |01>
multi_zero = np.kron(zero, zero)
multi_one = np.kron(zero, one)
print('Multiqubit computational dot product=', np.vdot(multi_zero, multi_one))

# our x basis states.
# zero = |+0> + |+1> 
# one = |+0> - |+1>
multi_x_zero = np.kron(x_zero, x_zero)
multi_x_one = np.kron(x_zero, x_one)
print('X basis dot product=', np.vdot(multi_x_zero, multi_x_one))

# our y basis states.
# zero = |+0> + i|+1>
# one = |+0> - i|+1>
multi_y_zero = np.kron(x_zero, y_zero)
multi_y_one = np.kron(x_zero, y_one)
print('Y basis dot product=', np.vdot(multi_y_zero, multi_y_one))

ident = np.identity(2)
multi_X0_trans = np.kron((cirq.Y**-0.5)._unitary_(), ident)
multi_Y0_trans = np.kron((cirq.X**0.5)._unitary_(), ident)
multi_X1_trans = np.kron(ident, (cirq.Y**-0.5)._unitary_())
multi_Y1_trans = np.kron(ident, (cirq.X**0.5)._unitary_())

print('Checking that X transform takes us from x states'
	  ' to computation states.')
print('Zero:', cirq.equal_up_to_global_phase(multi_X1_trans @ multi_X0_trans @ multi_x_zero, multi_zero))
print('One:', cirq.equal_up_to_global_phase(multi_X1_trans @ multi_X0_trans @ multi_x_one, multi_one))

print('Checking that Y transform takes us from y states'
	  ' to computation states.')
print('Zero:', cirq.equal_up_to_global_phase(multi_Y1_trans @ multi_X0_trans @ multi_y_zero, multi_zero))
print('One:', cirq.equal_up_to_global_phase(multi_Y1_trans @ multi_X0_trans @ multi_y_one, multi_one))
`
with output
`Multiqubit computational dot product= 0
X basis dot product= 0.0
Y basis dot product= 0j
Checking that X transform takes us from x states to computation states.
Zero: True
One: True
Checking that Y transform takes us from y states to computation states.
Zero: True
One: True
`
so I am convinced it's ok now